plugins/modules/terraform.py

```diff
@@ -473,6 +473,159 @@ def get_outputs(terraform_binary, project_path, state_file):
     return outputs
 
 
+# needs init
+def get_providers_schema(terraform_binary, project_path):
+    command = [
+        terraform_binary,
+        "providers",
+        "schema",
+        "-json",
+    ]  # in the command we have "providers schema", in the schema we have "provider_schemas"
+    rc, text, err = module.run_command(command, cwd=project_path)
+    if rc == 1:
+        module.warn(
+            "Could not get provider schemas. "
+            "\nstdout: {0}\nstderr: {1}".format(text, err)
+        )
+        provider_schemas = {}
+    elif rc != 0:
+        module.fail_json(
+            msg="Failure when getting provider schemas. "
+            "Exited {0}.\nstdout: {1}\nstderr: {2}".format(rc, text, err),
+            command=" ".join(command),
+        )
+    else:
+        provider_schemas = json.loads(text)
+    return provider_schemas
+
+
+def is_attribute_sensitive_in_providers_schema(provider_schemas, resource, attribute):
+    for provider_schema in provider_schemas[
+        "provider_schemas"
+    ]:
+        for resource_schema_key, resource_schema_value in (
+            provider_schemas["provider_schemas"]
+            .get(provider_schema)
+            .get("resource_schemas")
+            .items()
+        ):
+            if resource_schema_key == resource["type"]:
+                sensitive = resource_schema_value["block"]["attributes"][attribute].get(
+                    "sensitive", False
+                )  # sensitive exists only when it is True
+                return sensitive
+
+
+def is_attribute_in_sensitive_values(resource, attribute):
+    return attribute in resource["sensitive_values"]
+
+
+def filter_resource_attributes(state_file, provider_schemas):
+    # using .get() in case there is no existing .tfstate before apply
+    for resource in state_file.get("values", {}).get("root_module", {}).get("resources", {}):
+        attributes_to_remove = []
+        for attribute in resource["values"]:
+            if is_attribute_sensitive_in_providers_schema(
+                provider_schemas, resource, attribute
+            ):
+                attributes_to_remove.append(attribute)
+            elif is_attribute_in_sensitive_values(resource, attribute):
+                attributes_to_remove.append(attribute)
+        for attribute in attributes_to_remove:
+            resource["values"].pop(attribute)
```

Comment:
Please set values to `None` instead of popping the key, so the key is still present but without a value.

Reply:
Updated